coveo-testing/README.md

```diff
@@ -221,3 +221,456 @@ If you run `pytest --collect-only` you will obtain the following:
+The test needs to patch the function _in the tasks module_ since that's the context it will be called from. 
+Unfortunately, inspecting `get_api_client` from the `tasks` module at runtime leads us back to `mymodule.clients`.
+
+This single complexity means that hardcoding the context `mymodule.tasks` and symbol `get_api_client` into a string
+for the patch is the straightforward solution.
+
+But with `ref`, you specify the context separately:
+
+```python
+from unittest.mock import patch, MagicMock
+from coveo_testing.mocks import ref
+from mymodule.clients import get_api_client
+from mymodule.tasks import process
+
+
+@patch(*ref(get_api_client, context=process))
+def test(get_api_client_mock: MagicMock) -> None:
+    assert process() is None  # pretend this tests the process function
+```
+
+🚀 By giving a context to `ref`, the symbol `get_api_client` will be resolved from the context of `process`, which is the
+`mymodule.tasks` module. The result is `mymodule.tasks.get_api_client`.
+
+If either objects (`get_api_client` or `process`) are moved or renamed using a refactoring tool, the mock will still
+point to the correct name and context.
+
+🚀 And a nice bonus is that your IDE can jump to `get_api_client`'s definition from the test file now!
+
+It should be noted that this isn't just some string manipulation. `ref` will import and inspect modules and objects
+to make sure that they're correct. Here's a more complex case with a renamed symbol:
+
+The module:
+
+```python
+from typing import Optional
+from mymodule.clients import get_api_client as client_factory  # it got renamed! 😱
+
+def process() -> Optional[bool]:
+  client = client_factory()
+  return ...
+```
+
+The test:
+
+```python
+from unittest.mock import patch, MagicMock
+from coveo_testing.mocks import ref
+from mymodule.clients import get_api_client
+from mymodule.tasks import process
+
+
+@patch(*ref(get_api_client, context=process))
+def test(get_api_client_mock: MagicMock) -> None:
+    assert process() is None  # pretend this tests the process function
+```
+
+Notice how the test and patch did not change despite the renamed symbol?
+
+🚀 This is because `ref` will find `get_api_client` as `client_factory` when inspecting `mymodule.tasks` module,
+and return `mymodule.tasks.client_factory`.
+
+We can also use ref with `patch.object()` in order to patch a single instance. Consider the following code:
+
+```python
+from unittest.mock import patch
+from mymodule.clients import APIClient
+
+def test() -> None:
+  client = APIClient()
+  with patch.object(client, "_do_request"):
+    ...
+```
+
+🚀 By specifying `obj=True` to `ref`, you will obtain a `Tuple[instance, attribute_to_patch_as_a_string]` that you
+can unpack to `patch.object()`:
+
+```python
+from unittest.mock import patch
+from coveo_testing.mocks import ref
+from mymodule.clients import APIClient
+
+def test() -> None:
+  client = APIClient()
+  with patch.object(*ref(client._do_request, obj=True)):
+    ...
+```

```

Comment:
This is a pattern I'm seeing throughout the file. Some python code is indented with 2 spaces while other python code is indented with 4. I think that you should have them all be 4 spaces for ease of copy paste and for consistency.